src/py42/clients/_settings_managers.py

```diff
@@ -0,0 +1,520 @@
+from py42._internal.compat import str
+from py42._internal.compat import UserList
+from py42._internal.compat import UserDict
+from py42.exceptions import Py42Error
+
+
+def set_val(d, keys, value):
+    """Helper for setting nested values from a dict based on a list of keys."""
+    d = get_val(d, keys[:-1])
+    d[keys[-1]] = value
+
+
+def get_val(d, keys):
+    """Helper for getting nested values from a dict based on a list of keys."""
+    for key in keys:
+        d = d[key]
+    return d
+
+
+def bool_to_str(value):
+    if isinstance(value, bool):
+        return "true" if value else "false"
+    elif value in ("true", "false"):
+        return value
+    else:
+        raise ValueError("Value must be True or False")
+
+
+def str_to_bool(value):
+    if value == "true":
+        return True
+    if value == "false":
+        return False
+    else:
+        return value
+
+
+def days_to_minutes(days):
+    return str(int(days) * 1440)
+
+
+def minutes_to_days(minutes):
+    minutes = int(minutes)
+    return int(minutes / 1440)
+
+
+def show_change(val1, val2):
+    return "{} -> {}".format(val1, val2)
+
+
+def no_conversion(x):
+    return x
+
+
+class SettingProperty(object):
+    """Descriptor class to help manage changes to nested dict values. Assumes attributes
+    being managed are on a UserDict/UserList subclass.
+
+    Args:
+        name (str): name of attribute this class manages (changes will be registered with this name).
+        location (list): list of keys defining the location of the value being managed in the managed class.
+        get_converter (func, optional): function to convert retrieved values to preferred format.
+        set_converter (func, optional): function to convert values being set to preferred format.
+    """
+
+    def __init__(
+        self, name, location, get_converter=no_conversion, set_converter=no_conversion
+    ):
+        self.name = name
+        self.location = location
+        self.init_val = None
+        self.get_converter = get_converter
+        self.set_converter = set_converter
+
+    def __get__(self, instance, owner):
+        val = get_val(instance.data, self.location)
+        return self.get_converter(val)
+
+    def __set__(self, instance, val):
+        val = self.set_converter(val)
+        self._register_change(instance, val)
+        set_val(instance.data, self.location, val)
+
+    def _register_change(self, instance, new_val):
+        new_val = self.get_converter(new_val)
+        if self.init_val is None:
+            init_val = get_val(instance.data, self.location)
+            self.init_val = self.get_converter(init_val)
+        if self.init_val == new_val:
+            if self.name in instance.changes:
+                instance.changes.pop(self.name)
+        else:
+            instance.changes[self.name] = show_change(self.init_val, new_val)
+
+
+class TSettingProperty(object):
```

Comment:
What does t_settings mean? Type settings?

Reply:
It's for settings that get stored in the `t_settings` table of the database. The reason that there are two endpoints each for modifying devices/orgs is because there's a fundamental difference to the backend in how they're stored, so they have very different APIs for accessing and updating. I don't think there's an "official" name for these settings, they're commonly just referred to as t_settings.

The `/api/Org` and `/api/Device` endpoints can basically take the same json that you get from a GET request, and send that back with modifications in a PUT and it just works. However, the `/api/DeviceSettings` and `/api/OrgSettings` endpoints return results as a dict of keys like this:

```
  "data": {
    "device plan key": {
      "scope": "DEVICE",
      "value": "foo",
      "locked": false,
      "id": 2
    },
    "system key": {
      "scope": "SYSTEM",
      "value": "foo",
      "locked": false
    }
```
But require updates to be sent in a list of "packets" like this:

```
{
  "packets": [
    {
      "key": <key1>,
      "value": <value1>,
      "locked": <true/false>
    },
    {
      "key": <key2>,
      "value": <value2>,
      "locked": <true/false>
    },
    ...
  ]
}
```

So the `TSettingsProperty` is a helper that does the tracking and translation from the response data into packets suitable for posting. 

Since both `/api/OrgSettings` and `/api/DeviceSetting` both use the same table and API, I figured I'd name it after the technical backend they share... although I didn't implement any of the device t_settings in the device manager (since there's only a small handful that don't actually get changed much), so that shared-ness isn't clear from the existing code.